tests/MongoDB.AspNetCore.OData.Tests/TestServer.cs

```diff
@@ -0,0 +1,100 @@
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+using System;
+using System.Linq;
+using System.Net.Http;
+using System.Reflection;
+using System.Text.Json;
+using System.Threading.Tasks;
+using Json.Schema;
+using Microsoft.AspNetCore.Mvc.Testing;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+namespace MongoDB.AspNetCore.OData.Tests
+{
+    [TestClass]
+    internal static class TestServer
+    {
+        private static WebApplicationFactory<Program> __applicationFactory;
+        private static HttpClient __httpClient;
+
+        [AssemblyInitialize]
+        public static void TestServerInitialize(TestContext testContext)
+        {
+            __applicationFactory = new WebApplicationFactory<Program>();
+            __httpClient = __applicationFactory.CreateClient();
+        }
+
+        [AssemblyCleanup]
+        public static void TestServerInitializeCleanup()
+        {
+            __applicationFactory.Dispose();
+        }
+
+        public static async Task<JsonDocument> GetAndValidateODataRequestAsync(string requestUrl, string schemaName)
+        {
+            await using var responseStream = await __httpClient.GetStreamAsync(requestUrl);
+            var document = await JsonDocument.ParseAsync(responseStream);
+            ValidateResponseSchema(document, schemaName);
+
+            return document;
+        }
+
+        public static async Task<JsonDocument> GetAndValidateODataRequestAsync<TItem>(string requestUrl, string schemaName, Func<TItem, bool> itemValidator)
+        {
+            var document = await GetAndValidateODataRequestAsync(requestUrl, schemaName);
+
+            var valueNode = document.RootElement.GetProperty("value");
+            if (valueNode.GetArrayLength() == 0)
+            {
+                Assert.Fail("Request result should not be empty");
+            }
+
+            foreach (var itemNode in valueNode.EnumerateArray())
+            {
+                var item = itemNode.Deserialize<TItem>();
+                Assert.IsTrue(itemValidator(item), $"Response item does not match validation criteria: {itemNode}");
+            }
+
+            return document;
+        }
+
+        private static JsonSchema LoadSchema(string schemaName)
+        {
+            var assembly = Assembly.GetExecutingAssembly();
+            using var schemaStream = assembly.GetManifestResourceStream($"MongoDB.AspNetCore.OData.Tests.Schemas.{schemaName}.json");
+
+            return JsonSchema.FromStream(schemaStream).GetAwaiter().GetResult();
+        }
+
+        private static void ValidateResponseSchema(JsonDocument document, string schemaName)
```

Comment:
Consider making this async, so `LoadSchema` can be async and can get rid of `.GetAwaiter().GetResult()`.

Also is the same schema is loaded multiple times? If so, maybe a simple cache?

Reply:
Done